File: compress/src/java/compresslib/Deflate.java
package compresslib;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * DEFLATE 부호기 — SPEC §10.
 *
 * <p>RFC 가 부호기에 맡긴 선택을 전부 못 박은 것이 이 파일이다. 블록
 * 65535, 값을 정확히 세어 가장 작은 것, 같으면 stored → fixed →
 * dynamic. 값을 어림하면 언어마다 반올림이 달라져 블록 종류가 갈린다.
 */
public final class Deflate {
  public static final int BLOCK_SIZE = 65535;
  public static final int HASH_BITS = 15;
  public static final int HASH_SIZE = 1 << HASH_BITS;
  public static final int CHAIN_LIMIT = 128;
  static final int NIL = -1;

  private Deflate() {}

  private static final class ClItem {
    final int sym;
    final int value;
    final int nbits;

    ClItem(int sym, int value, int nbits) {
      this.sym = sym;
      this.value = value;
      this.nbits = nbits;
    }
  }

  static int hash3(byte[] s, int i) {
    int h = ((s[i] & 0xFF) << 10) ^ ((s[i + 1] & 0xFF) << 5)
        ^ (s[i + 2] & 0xFF);
    return h & (HASH_SIZE - 1);
  }

  public static List<Lzss.Token> parse(byte[] src) {
    int n = src.length;
    int[] head = new int[HASH_SIZE];
    Arrays.fill(head, NIL);
    int[] prev = new int[Math.max(n, 1)];
    Arrays.fill(prev, NIL);
    List<Lzss.Token> tokens = new ArrayList<>();
    int[] best = new int[2];      // {길이, 거리}
    int i = 0;
    while (i < n) {
      find(src, i, head, prev, best);
      int ln = best[0];
      int dist = best[1];
      insert(src, i, head, prev);
      if (ln >= DeflateTables.MIN_MATCH) {
        int nxtLen = 0;
        if (i + 1 < n) {
          find(src, i + 1, head, prev, best);
          nxtLen = best[0];
        }
        if (nxtLen > ln) {      // 게으른 일치
          tokens.add(new Lzss.Token(false, src[i] & 0xFF, 0, 0));
          i++;
          continue;
        }
        for (int k = 1; k < ln; k++) {
          insert(src, i + k, head, prev);
        }
        tokens.add(new Lzss.Token(true, 0, ln, dist));
        i += ln;
      } else {
        tokens.add(new Lzss.Token(false, src[i] & 0xFF, 0, 0));
        i++;
      }
    }
    return tokens;
  }

  private static void insert(byte[] src, int p, int[] head,
      int[] prev) {
    if (p + DeflateTables.MIN_MATCH <= src.length) {
      int h = hash3(src, p);
      prev[p] = head[h];
      head[h] = p;
    }
  }

  private static void find(byte[] src, int p, int[] head, int[] prev,
      int[] out) {
    out[0] = 0;
    out[1] = 0;
    int n = src.length;
    if (p + DeflateTables.MIN_MATCH > n) {
      return;
    }
    int limit = Math.min(DeflateTables.MAX_MATCH, n - p);
    int cand = head[hash3(src, p)];
    int probes = 0;
    while (cand != NIL && probes < CHAIN_LIMIT) {
      int dist = p - cand;
      if (dist > DeflateTables.MAX_DIST) {
        break;
      }
      int ln = 0;
      while (ln < limit && src[cand + ln] == src[p + ln]) {
        ln++;
      }
      if (ln > out[0]) {
        out[0] = ln;
        out[1] = dist;
        if (ln == limit) {
          break;
        }
      }
      cand = prev[cand];
      probes++;
    }
  }

  private static final class Block {
    final int tokA;
    final int tokB;
    final int inA;
    final int inB;

    Block(int tokA, int tokB, int inA, int inB) {
      this.tokA = tokA;
      this.tokB = tokB;
      this.inA = inA;
      this.inB = inB;
    }
  }

  private static List<Block> splitBlocks(List<Lzss.Token> tokens) {
    List<Block> blocks = new ArrayList<>();
    int tokStart = 0;
    int inStart = 0;
    int cur = 0;
    for (int k = 0; k < tokens.size(); k++) {
      Lzss.Token t = tokens.get(k);
      cur += t.isMatch ? t.length : 1;
      if (cur >= BLOCK_SIZE) {
        blocks.add(new Block(tokStart, k + 1, inStart, inStart + cur));
        tokStart = k + 1;
        inStart += cur;
        cur = 0;
      }
    }
    if (cur > 0 || blocks.isEmpty()) {
      blocks.add(
          new Block(tokStart, tokens.size(), inStart, inStart + cur));
    }
    return blocks;
  }

  /** 빈도를 세고 여분 비트 수를 돌려준다. */
  private static int freqsOf(List<Lzss.Token> tokens, int a, int b,
      long[] lit, long[] dst) {
    Arrays.fill(lit, 0);
    Arrays.fill(dst, 0);
    int extra = 0;
    for (int k = a; k < b; k++) {
      Lzss.Token t = tokens.get(k);
      if (t.isMatch) {
        int code = DeflateTables.LENGTH_CODE[t.length];
        lit[code]++;
        extra += DeflateTables.LENGTH_EXTRA[code - 257];
        int dc = DeflateTables.distCode(t.dist);
        dst[dc]++;
        extra += DeflateTables.DIST_EXTRA[dc];
      } else {
        lit[t.literal]++;
      }
    }
    lit[DeflateTables.END_OF_BLOCK]++;
    return extra;
  }

  private static long bodyBits(long[] lit, long[] dst, int extra,
      int[] litLen, int[] dstLen) {
    long bits = extra;
    for (int s = 0; s < lit.length; s++) {
      if (lit[s] > 0) {
        bits += lit[s] * litLen[s];
      }
    }
    for (int s = 0; s < dst.length; s++) {
      if (dst[s] > 0) {
        bits += dst[s] * dstLen[s];
      }
    }
    return bits;
  }

  /** 부호 길이 배열 → (기호, 여분 값, 여분 비트). 왼쪽부터 탐욕. */
  private static List<ClItem> clEncode(int[] lengths) {
    List<ClItem> out = new ArrayList<>();
    int i = 0;
    int n = lengths.length;
    while (i < n) {
      int cur = lengths[i];
      int run = 1;
      while (i + run < n && lengths[i + run] == cur) {
        run++;
      }
      if (cur == 0) {
        while (run >= 3) {
          int k;
          if (run >= 11) {
            k = Math.min(run, 138);
            out.add(new ClItem(DeflateTables.CL_ZERO_LONG, k - 11, 7));
          } else {
            k = Math.min(run, 10);
            out.add(new ClItem(DeflateTables.CL_ZERO_SHORT, k - 3, 3));
          }
          run -= k;
          i += k;
        }
        for (int j = 0; j < run; j++) {
          out.add(new ClItem(0, 0, 0));
          i++;
        }
      } else {
        out.add(new ClItem(cur, 0, 0));
        i++;
        run--;
        while (run >= 3) {
          int k = Math.min(run, 6);
          out.add(new ClItem(DeflateTables.CL_REPEAT, k - 3, 2));
          run -= k;
          i += k;
        }
        for (int j = 0; j < run; j++) {
          out.add(new ClItem(cur, 0, 0));
          i++;
        }
      }
    }
    return out;
  }

  private static int lastUsed(int[] lengths) {
    for (int s = lengths.length; s > 0; s--) {
      if (lengths[s - 1] > 0) {
        return s;
      }
    }
    return 0;
  }

  private static final class DynamicPlan {
    final int[] litLen;
    final int[] dstLen;
    final int[] clLen;
    final List<ClItem> items;
    final int hlit;
    final int hdist;
    int hclen;
    final long bits;

    DynamicPlan(long[] litFreq, long[] dstFreq, int extra) {
      litLen = Huffman.codeLengths(litFreq, Huffman.MAX_LENGTH);
      dstLen = Huffman.codeLengths(dstFreq, Huffman.MAX_LENGTH);
      boolean any = false;
      for (int l : dstLen) {
        any = any || l > 0;
      }
      if (!any) {
        // 일치가 하나도 없는 블록. 거리 부호를 안 보낼 수는 없으므로
        // 하나를 길이 1 로 보낸다 — 쓰이지 않는 부호다 (§10.5).
        dstLen[0] = 1;
      }
      hlit = Math.max(257, lastUsed(litLen));
      hdist = Math.max(1, lastUsed(dstLen));
      int[] joined = new int[hlit + hdist];
      System.arraycopy(litLen, 0, joined, 0, hlit);
      System.arraycopy(dstLen, 0, joined, hlit, hdist);
      items = clEncode(joined);
      long[] clFreq = new long[DeflateTables.CL_SYMBOLS];
      for (ClItem it : items) {
        clFreq[it.sym]++;
      }
      clLen = Huffman.codeLengths(clFreq, DeflateTables.CL_MAX_LENGTH);
      hclen = DeflateTables.CL_SYMBOLS;
      while (hclen > 4
          && clLen[DeflateTables.CL_ORDER[hclen - 1]] == 0) {
        hclen--;
      }
      long header = 5 + 5 + 4 + 3L * hclen;
      for (ClItem it : items) {
        header += clLen[it.sym] + it.nbits;
      }
      bits = 3 + header
          + bodyBits(litFreq, dstFreq, extra, litLen, dstLen);
    }
  }

  private static void writeBody(BitIO.LsbWriter w,
      List<Lzss.Token> tokens, int a, int b, int[] litLen,
      int[] litCode, int[] dstLen, int[] dstCode) {
    for (int k = a; k < b; k++) {
      Lzss.Token t = tokens.get(k);
      if (t.isMatch) {
        int code = DeflateTables.LENGTH_CODE[t.length];
        w.writeCode(litCode[code], litLen[code]);
        int idx = code - 257;
        if (DeflateTables.LENGTH_EXTRA[idx] > 0) {
          w.writeBits(t.length - DeflateTables.LENGTH_BASE[idx],
              DeflateTables.LENGTH_EXTRA[idx]);
        }
        int dc = DeflateTables.distCode(t.dist);
        w.writeCode(dstCode[dc], dstLen[dc]);
        if (DeflateTables.DIST_EXTRA[dc] > 0) {
          w.writeBits(t.dist - DeflateTables.DIST_BASE[dc],
              DeflateTables.DIST_EXTRA[dc]);
        }
      } else {
        w.writeCode(litCode[t.literal], litLen[t.literal]);
      }
    }
    int eob = DeflateTables.END_OF_BLOCK;
    w.writeCode(litCode[eob], litLen[eob]);
  }

  public static byte[] deflateRaw(byte[] src) {
    BitIO.LsbWriter w = new BitIO.LsbWriter();
    if (src.length == 0) {
      // 마지막 고정 블록 하나, 안에는 블록 끝 기호뿐. 두 바이트 03 00.
      w.writeBits(1, 1);
      w.writeBits(1, 2);
      w.writeCode(0, 7);
      w.flush();
      return w.bytes();
    }
    List<Lzss.Token> tokens = parse(src);
    List<Block> blocks = splitBlocks(tokens);
    int[] fixedCode =
        Huffman.canonicalCodes(DeflateTables.FIXED_LITLEN);
    int[] fixedDcode = Huffman.canonicalCodes(DeflateTables.FIXED_DIST);
    long[] litFreq = new long[DeflateTables.LITLEN_SYMBOLS];
    long[] dstFreq = new long[DeflateTables.DIST_SYMBOLS];
    for (int k = 0; k < blocks.size(); k++) {
      Block blk = blocks.get(k);
      int last = (k == blocks.size() - 1) ? 1 : 0;
      int extra = freqsOf(tokens, blk.tokA, blk.tokB, litFreq, dstFreq);
      int rawLen = blk.inB - blk.inA;
      // stored 의 값은 지금 비트 자리에 달려 있다 — 정렬 때문이다.
      int pad = (8 - ((w.bitPos() + 3) % 8)) % 8;
      long costStored = 3L + pad + 32 + 8L * rawLen;
      long costFixed = 3 + bodyBits(litFreq, dstFreq, extra,
          DeflateTables.FIXED_LITLEN, DeflateTables.FIXED_DIST);
      DynamicPlan plan = new DynamicPlan(litFreq, dstFreq, extra);
      if (costStored <= costFixed && costStored <= plan.bits) {
        w.writeBits(last, 1);
        w.writeBits(0, 2);
        w.align();
        w.writeBits(rawLen, 16);
        w.writeBits(rawLen ^ 0xFFFF, 16);
        for (int j = blk.inA; j < blk.inB; j++) {
          w.writeBits(src[j] & 0xFF, 8);
        }
        continue;
      }
      w.writeBits(last, 1);
      if (costFixed <= plan.bits) {
        w.writeBits(1, 2);
        writeBody(w, tokens, blk.tokA, blk.tokB,
            DeflateTables.FIXED_LITLEN, fixedCode,
            DeflateTables.FIXED_DIST, fixedDcode);
        continue;
      }
      w.writeBits(2, 2);
      w.writeBits(plan.hlit - 257, 5);
      w.writeBits(plan.hdist - 1, 5);
      w.writeBits(plan.hclen - 4, 4);
      for (int i = 0; i < plan.hclen; i++) {
        w.writeBits(plan.clLen[DeflateTables.CL_ORDER[i]], 3);
      }
      int[] clCode = Huffman.canonicalCodes(plan.clLen);
      for (ClItem it : plan.items) {
        w.writeCode(clCode[it.sym], plan.clLen[it.sym]);
        if (it.nbits > 0) {
          w.writeBits(it.value, it.nbits);
        }
      }
      writeBody(w, tokens, blk.tokA, blk.tokB, plan.litLen,
          Huffman.canonicalCodes(plan.litLen), plan.dstLen,
          Huffman.canonicalCodes(plan.dstLen));
    }
    w.flush();
    return w.bytes();
  }

  // 골든 코덱. 다른 모듈과 달리 varint 헤더가 없다 — 남의 형식이라
  // 우리가 얹을 자리가 없고, 원본 길이는 스트림 자신이 알고 있다.
  public static byte[] encode(byte[] src) {
    return deflateRaw(src);
  }

  public static byte[] decode(byte[] src) {
    return Inflate.inflateRaw(src);
  }
}

File: compress/src/py/compresslib/deflate.py
# -*- coding: utf-8 -*-
"""DEFLATE 부호기 — SPEC §10.

RFC 1951 이 부호기에 맡긴 선택을 전부 못 박은 것이 이 파일이다. 형식은
남의 것이지만 **어떤 파일을 낼지** 는 우리가 정한다. 그 선택이 정해져
있어야 다섯 언어가 같은 바이트를 낸다.

정한 것:
  · 블록 65535 바이트 (stored 의 LEN 이 16비트라 65536 은 못 담는다)
  · 세 종류의 값을 정확히 세어 가장 작은 것. 같으면 stored→fixed→dynamic
  · 일치 찾기는 해시 사슬 128번 + 게으른 일치 (§10.4)
  · gzip 헤더의 MTIME 은 0 — 진짜 gzip 과 달리 우리 출력은 재현된다

zlib 에 있는 good_match·nice_length 같은 조절기는 안 둔다. 그것들이 zlib
출력을 압축 수준에 따라 달라지게 하는 물건이고, 우리는 한 가지 행동만
한다. 진짜 gzip -9 와의 비율 차이는 9부에서 캡처로 견준다.

찾기는 위치마다 최대 128번 × 최대 258바이트 비교다.
"""
from compresslib import bitio, huffman, inflate
from compresslib import deflate_tables as T

BLOCK_SIZE = 65535
HASH_BITS = 15
HASH_SIZE = 1 << HASH_BITS
CHAIN_LIMIT = 128
NIL = -1

STORED, FIXED, DYNAMIC = 0, 1, 2


# -------------------------------------------------------------------
# 파싱
def _hash3(src, i):
    h = (src[i] << 10) ^ (src[i + 1] << 5) ^ src[i + 2]
    return h & (HASH_SIZE - 1)


def parse(src):
    """리터럴과 (길이, 거리) 의 열. 게으른 일치까지 여기서 끝낸다."""
    n = len(src)
    head = [NIL] * HASH_SIZE
    prev = [NIL] * max(n, 1)
    tokens = []

    def insert(p):
        if p + T.MIN_MATCH <= n:
            h = _hash3(src, p)
            prev[p] = head[h]
            head[h] = p

    def find(p):
        if p + T.MIN_MATCH > n:
            return 0, 0
        best_len = best_dist = 0
        limit = min(T.MAX_MATCH, n - p)
        cand = head[_hash3(src, p)]
        probes = 0
        while cand != NIL and probes < CHAIN_LIMIT:
            dist = p - cand
            if dist > T.MAX_DIST:
                break
            ln = 0
            while ln < limit and src[cand + ln] == src[p + ln]:
                ln += 1
            if ln > best_len:
                best_len, best_dist = ln, dist
                if ln == limit:
                    break
            cand = prev[cand]
            probes += 1
        return best_len, best_dist

    i = 0
    while i < n:
        ln, dist = find(i)
        insert(i)
        if ln >= T.MIN_MATCH:
            # 게으른 일치 — 한 칸 뒤에서 **더 긴** 것이 있으면 이번은
            # 버린다.
            nxt_len = find(i + 1)[0] if i + 1 < n else 0
            if nxt_len > ln:
                tokens.append(src[i])
                i += 1
                continue
            for k in range(1, ln):
                insert(i + k)
            tokens.append((ln, dist))
            i += ln
        else:
            tokens.append(src[i])
            i += 1
    return tokens


def split_blocks(tokens):
    """(토큰 시작, 토큰 끝, 입력 시작, 입력 끝) 목록."""
    blocks = []
    tok_start, in_start, cur = 0, 0, 0
    for k, t in enumerate(tokens):
        cur += t[0] if isinstance(t, tuple) else 1
        if cur >= BLOCK_SIZE:
            blocks.append((tok_start, k + 1, in_start, in_start + cur))
            tok_start, in_start, cur = k + 1, in_start + cur, 0
    if cur or not blocks:
        blocks.append((tok_start, len(tokens),
                       in_start, in_start + cur))
    return blocks


# ------------------------------------------------------------- 값 계산
def _freqs(tokens):
    lit = [0] * T.LITLEN_SYMBOLS
    dst = [0] * T.DIST_SYMBOLS
    extra = 0
    for t in tokens:
        if isinstance(t, tuple):
            ln, d = t
            code = T.LENGTH_CODE[ln]
            lit[code] += 1
            extra += T.LENGTH_EXTRA[code - 257]
            dc = T.dist_code(d)
            dst[dc] += 1
            extra += T.DIST_EXTRA[dc]
        else:
            lit[t] += 1
    lit[T.END_OF_BLOCK] += 1
    return lit, dst, extra


def _body_bits(lit_freq, dst_freq, extra, lit_len, dst_len):
    bits = extra
    for s, f in enumerate(lit_freq):
        if f:
            bits += f * lit_len[s]
    for s, f in enumerate(dst_freq):
        if f:
            bits += f * dst_len[s]
    return bits


def cl_encode(lengths):
    """부호 길이 배열 → (기호, 여분 값, 여분 비트). 왼쪽부터 탐욕."""
    out = []
    i, n = 0, len(lengths)
    while i < n:
        cur = lengths[i]
        run = 1
        while i + run < n and lengths[i + run] == cur:
            run += 1
        if cur == 0:
            while run >= 3:
                if run >= 11:
                    k = min(run, 138)
                    out.append((T.CL_ZERO_LONG, k - 11, 7))
                else:
                    k = min(run, 10)
                    out.append((T.CL_ZERO_SHORT, k - 3, 3))
                run -= k
                i += k
            for _ in range(run):
                out.append((0, 0, 0))
                i += 1
        else:
            out.append((cur, 0, 0))
            i += 1
            run -= 1
            while run >= 3:
                k = min(run, 6)
                out.append((T.CL_REPEAT, k - 3, 2))
                run -= k
                i += k
            for _ in range(run):
                out.append((cur, 0, 0))
                i += 1
    return out


class DynamicPlan:
    """동적 블록 하나를 쓰는 데 필요한 것 전부와 그 비트 수."""

    def __init__(self, lit_freq, dst_freq, extra):
        lim = huffman.MAX_LENGTH
        self.lit_len = huffman.code_lengths(lit_freq, lim)
        self.dst_len = huffman.code_lengths(dst_freq, lim)
        if not any(self.dst_len):
            # 일치가 하나도 없는 블록. 거리 부호를 안 보낼 수는 없으므로
            # 하나를 길이 1 로 보낸다 — 쓰이지 않는 부호다 (§10.5).
            self.dst_len[0] = 1
        self.hlit = max(257, _last_used(self.lit_len))
        self.hdist = max(1, _last_used(self.dst_len))
        self.items = cl_encode(self.lit_len[:self.hlit]
                               + self.dst_len[:self.hdist])
        cl_freq = [0] * T.CL_SYMBOLS
        for sym, _v, _n in self.items:
            cl_freq[sym] += 1
        self.cl_len = huffman.code_lengths(cl_freq, T.CL_MAX_LENGTH)
        self.hclen = T.CL_SYMBOLS
        while (self.hclen > 4
               and self.cl_len[T.CL_ORDER[self.hclen - 1]] == 0):
            self.hclen -= 1
        header = 5 + 5 + 4 + 3 * self.hclen
        for sym, _v, nbits in self.items:
            header += self.cl_len[sym] + nbits
        self.bits = 3 + header + _body_bits(lit_freq, dst_freq, extra,
                                            self.lit_len, self.dst_len)


def _last_used(lengths):
    for s in range(len(lengths) - 1, -1, -1):
        if lengths[s]:
            return s + 1
    return 0


# --------------------------------------------------------------
# 내보내기
def _write_body(w, tokens, lit_len, lit_code, dst_len, dst_code):
    for t in tokens:
        if isinstance(t, tuple):
            ln, d = t
            code = T.LENGTH_CODE[ln]
            w.write_code(lit_code[code], lit_len[code])
            idx = code - 257
            if T.LENGTH_EXTRA[idx]:
                w.write_bits(ln - T.LENGTH_BASE[idx],
                             T.LENGTH_EXTRA[idx])
            dc = T.dist_code(d)
            w.write_code(dst_code[dc], dst_len[dc])
            if T.DIST_EXTRA[dc]:
                w.write_bits(d - T.DIST_BASE[dc], T.DIST_EXTRA[dc])
        else:
            w.write_code(lit_code[t], lit_len[t])
    w.write_code(lit_code[T.END_OF_BLOCK], lit_len[T.END_OF_BLOCK])


def _write_stored(w, data, final):
    w.write_bits(final, 1)
    w.write_bits(STORED, 2)
    w.align()
    w.write_bits(len(data), 16)
    w.write_bits(len(data) ^ 0xFFFF, 16)
    for b in data:
        w.write_bits(b, 8)


def deflate_raw(src):
    w = bitio.LsbWriter()
    if not src:
        # 마지막 고정 블록 하나, 안에는 블록 끝 기호뿐. 두 바이트 03 00.
        w.write_bits(1, 1)
        w.write_bits(FIXED, 2)
        w.write_code(0, 7)
        w.flush()
        return w.bytes()

    tokens = parse(src)
    blocks = split_blocks(tokens)
    fixed_code = huffman.canonical_codes(T.FIXED_LITLEN)
    fixed_dcode = huffman.canonical_codes(T.FIXED_DIST)
    for k, (ta, tb, ia, ib) in enumerate(blocks):
        final = 1 if k == len(blocks) - 1 else 0
        part = tokens[ta:tb]
        lit_freq, dst_freq, extra = _freqs(part)
        raw = src[ia:ib]
        # stored 의 값은 지금 비트 자리에 달려 있다 — 정렬 때문이다.
        pad = (-(w.bit_pos() + 3)) % 8
        cost_stored = 3 + pad + 32 + 8 * len(raw)
        cost_fixed = 3 + _body_bits(lit_freq, dst_freq, extra,
                                    T.FIXED_LITLEN, T.FIXED_DIST)
        plan = DynamicPlan(lit_freq, dst_freq, extra)
        # 같으면 stored → fixed → dynamic 순. 값을 어림하면 언어마다
        # 반올림이 달라져 블록 종류가 갈린다. 그래서 정확히 센다.
        if cost_stored <= cost_fixed and cost_stored <= plan.bits:
            _write_stored(w, raw, final)
            continue
        w.write_bits(final, 1)
        if cost_fixed <= plan.bits:
            w.write_bits(FIXED, 2)
            _write_body(w, part, T.FIXED_LITLEN, fixed_code,
                        T.FIXED_DIST, fixed_dcode)
            continue
        w.write_bits(DYNAMIC, 2)
        w.write_bits(plan.hlit - 257, 5)
        w.write_bits(plan.hdist - 1, 5)
        w.write_bits(plan.hclen - 4, 4)
        for i in range(plan.hclen):
            w.write_bits(plan.cl_len[T.CL_ORDER[i]], 3)
        cl_code = huffman.canonical_codes(plan.cl_len)
        for sym, val, nbits in plan.items:
            w.write_code(cl_code[sym], plan.cl_len[sym])
            if nbits:
                w.write_bits(val, nbits)
        _write_body(w, part, plan.lit_len,
                    huffman.canonical_codes(plan.lit_len),
                    plan.dst_len, huffman.canonical_codes(plan.dst_len))
    w.flush()
    return w.bytes()


inflate_raw = inflate.inflate_raw


# ------------------------------------------------------------ 골든 코덱
# 다른 모듈과 달리 varint 헤더가 없다. 남의 형식이라 우리가 얹을 자리가
# 없기 때문이다 — 원본 길이는 스트림 자신이 알고 있다.
def encode(src):
    return deflate_raw(src)


def decode(src):
    return inflate.inflate_raw(src)
